File: packages-js/bpmn-flowaudit/src/enrichment/citations.ts
/**
 * Recognising legal citations in free text and resolving EU acts (CELEX,
 * ELI, EUR-Lex URL) – purely syntactic, no network access.
 *
 * Same patterns as `auditcore_bpmn.citations`: long form („Artikel 73
 * Absatz 2 Buchstabe b der Verordnung (EU) 2021/1060“), short form („Art. 74
 * Abs. 2 UAbs. 2“), sections („§ 44 Absatz 1 LHO“, „§§ 23 und 44 LHO“),
 * administrative rules („VV Nummer 4.2 zu § 44 LHO“) and annexes.
 */

import { actLong, citation, normalized } from '../model/legalBasis'
import type { LegalBasis } from '../schema/types'

export type CitationForm = 'vv' | 'long' | 'short' | 'sections' | 'section'

export interface CitationHit {
  legalBasis: LegalBasis
  start: number
  end: number
  text: string
  form: CitationForm
}

const EU_ACT = String.raw`(?:(?:Delegierten|Durchführungs)\s*)?(?:Verordnung|Richtlinie|VO|RL)\s*\((?:EU|EG|EWG|EU,\s*Euratom|EG,\s*Euratom)\)\s*(?:Nr\.\s*)?\d{1,4}\/\d{1,4}`
const SUB_LONG = String.raw`(?:\s+Absatz\s+(?<abs>\d+[a-z]?))?(?:\s+Unterabsatz\s+(?<uabs>\d+))?(?:\s+Satz\s+(?<satz>\d+))?(?:\s+Buchstabe\s+(?<buchst>[a-z]{1,3}))?(?:\s+(?:Nummer|Ziffer)\s+(?<nr>[\divx.]+))?`
const SUB_SHORT = String.raw`(?:\s*Abs\.\s*(?<abs>\d+[a-z]?))?(?:\s*UAbs\.\s*(?<uabs>\d+))?(?:\s*(?:S\.|Satz)\s*(?<satz>\d+))?(?:\s*(?:Buchst\.|lit\.)\s*(?<buchst>[a-z]{1,3}))?(?:\s*(?:Nr\.|Ziff\.)\s*(?<nr>[\divx.]+))?`
const SUB_SECTION = String.raw`(?:\s+(?:Absatz|Abs\.)\s+(?<abs>\d+[a-z]?))?(?:\s+Unterabsatz\s+(?<uabs>\d+))?(?:\s+(?:Satz|S\.)\s+(?<satz>\d+))?(?:\s+Buchstabe\s+(?<buchst>[a-z]{1,3}))?(?:\s+(?:Nummer|Nr\.|Ziffer)\s+(?<nr>[\divx.]+))?`
const NATIONAL_ACT = String.raw`(?<norm>(?:[A-ZÄÖÜ][A-Za-zÄÖÜäöüß]*[A-ZÄÖÜ][A-Za-zÄÖÜäöüß]*|UVgO|VgV|GWB))(?![A-Za-zÄÖÜäöüß0-9_])`

const LONG = new RegExp(String.raw`(?<kopf>Artikel\s+(?<art>\d+[a-z]?)|Anhang\s+(?<anh>[IVXLC]+))${SUB_LONG}\s+(?:der|des)\s+(?<norm>${EU_ACT})`, 'g')
const SHORT = new RegExp(String.raw`Art\.\s*(?<art>\d+[a-z]?)${SUB_SHORT}(?:\s+(?<norm>${EU_ACT}|CPR|Dach-VO))?`, 'g')
const SECTION = new RegExp(String.raw`§\s*(?<par>\d+[a-z]?)${SUB_SECTION}\s+${NATIONAL_ACT}`, 'g')
const SECTIONS = new RegExp(String.raw`§§\s*(?<liste>\d+[a-z]?(?:\s*(?:,|und|bis)\s*\d+[a-z]?)+)\s+${NATIONAL_ACT}`, 'g')
const VV = new RegExp(String.raw`VV\s+(?:Nummer|Nr\.)\s*(?<nr>[\d.]+)\s+zu\s+§\s*(?<par>\d+[a-z]?)\s+${NATIONAL_ACT}`, 'g')

type Groups = Record<string, string | undefined>

function group(groups: Groups, name: string): string | undefined {
  const value = groups[name]?.trim()
  return value ? value : undefined
}

function subdivisions(groups: Groups): Partial<LegalBasis> {
  return {
    paragraph: group(groups, 'abs'),
    subparagraph: group(groups, 'uabs'),
    sentence: group(groups, 'satz'),
    point: group(groups, 'buchst'),
    number: group(groups, 'nr'),
  }
}

function compact(basis: LegalBasis): LegalBasis {
  return Object.fromEntries(Object.entries(basis).filter(([, value]) => value !== undefined)) as LegalBasis
}

interface Recogniser {
  pattern: RegExp
  form: CitationForm
  build: (groups: Groups) => LegalBasis[]
}

const RECOGNISERS: Recogniser[] = [
  { pattern: VV, form: 'vv', build: (g) => [{ act: `VV zu § ${g.par} ${g.norm}`, number: g.nr }] },
  {
    pattern: LONG,
    form: 'long',
    build: (g) => [{ act: g.norm, article: group(g, 'art'), annex: group(g, 'anh'), ...subdivisions(g) }],
  },
  { pattern: SHORT, form: 'short', build: (g) => [{ act: group(g, 'norm'), article: g.art, ...subdivisions(g) }] },
  {
    pattern: SECTIONS,
    form: 'sections',
    build: (g) => (g.liste?.match(/\d+[a-z]?/g) ?? []).map((section) => ({ act: g.norm, section })),
  },
  { pattern: SECTION, form: 'section', build: (g) => [{ act: g.norm, section: g.par, ...subdivisions(g) }] },
]

/** Finds legal citations in free text (e.g. `bpmn:documentation`), without overlaps. */
export function findCitations(text: string | null | undefined): CitationHit[] {
  if (!text) return []
  const hits: CitationHit[] = []
  const taken: [number, number][] = []
  const free = (start: number, end: number) => taken.every(([s, e]) => end <= s || start >= e)
  for (const { pattern, form, build } of RECOGNISERS) {
    for (const match of text.matchAll(pattern)) {
      const start = match.index ?? 0
      const end = start + match[0].length
      if (!free(start, end)) continue
      taken.push([start, end])
      for (const basis of build((match.groups ?? {}) as Groups)) {
        hits.push({ legalBasis: normalized(compact(basis)), start, end, text: match[0], form })
      }
    }
  }
  return hits.sort((a, b) => a.start - b.start || citation(a.legalBasis).localeCompare(citation(b.legalBasis)))
}

/** Reads a single citation; without a hit it stays free text. */
export function parseCitation(text: string): LegalBasis {
  const hits = findCitations(text)
  const [only] = hits
  if (hits.length === 1 && only && only.text.trim() === text.trim()) return only.legalBasis
  return text.trim() ? { text: text.trim() } : {}
}

// ---------------------------------------------------------------------------
// EU acts
// ---------------------------------------------------------------------------

const EU_ACT_PARTS = /(?<typ>Delegierte(?:n)? Verordnung|Durchführungsverordnung|Verordnung|Richtlinie|Beschluss|Entscheidung)\s*\((?<org>EU|EG|EWG|EU,\s*Euratom|EG,\s*Euratom)\)\s*(?<nr>Nr\.\s*)?(?<a>\d{1,4})\/(?<b>\d{1,4})/
const OLD_DIRECTIVE = /Richtlinie\s+(?<a>\d{4})\/(?<b>\d{1,4})\/(?<org>EU|EG|EWG)/
const ELI_KIND: Record<string, string> = {
  Verordnung: 'reg',
  'Delegierte Verordnung': 'reg_del',
  'Delegierten Verordnung': 'reg_del',
  Durchführungsverordnung: 'reg_impl',
  Richtlinie: 'dir',
  Beschluss: 'dec',
  Entscheidung: 'dec',
}
const CELEX_TYPE: Record<string, string> = { reg: 'R', reg_del: 'R', reg_impl: 'R', dir: 'L', dec: 'D' }

const isYear = (value: number) => value >= 1950 && value <= 2100

/** `[eliKind, year, number]` of an EU act, or `null`. */
export function euAct(act: string): [string, number, number] | null {
  const text = actLong(act)
  const match = EU_ACT_PARTS.exec(text)
  if (match?.groups) {
    const kind = ELI_KIND[match.groups.typ ?? '']
    if (!kind) return null
    const first = Number(match.groups.a)
    const second = Number(match.groups.b)
    const numberFirst = kind !== 'dir' && (Boolean(match.groups.nr) || (isYear(second) && !isYear(first)))
    const [year, number] = numberFirst ? [second, first] : [first, second]
    return isYear(year) ? [kind, year, number] : null
  }
  const old = OLD_DIRECTIVE.exec(text)
  return old?.groups ? ['dir', Number(old.groups.a), Number(old.groups.b)] : null
}

/** Fills missing CELEX/ELI/URL of an EU act; existing values stay. */
export function completeEuAct(basis: LegalBasis): LegalBasis {
  const found = basis.act ? euAct(basis.act) : null
  if (!found) return basis
  const [kind, year, number] = found
  const celex = `3${year}${CELEX_TYPE[kind]}${String(number).padStart(4, '0')}`
  return {
    ...basis,
    celex: basis.celex ?? celex,
    eli: basis.eli ?? `http://data.europa.eu/eli/${kind}/${year}/${number}/oj`,
    url: basis.url ?? `https://eur-lex.europa.eu/legal-content/DE/TXT/?uri=CELEX:${celex}`,
  }
}

File: packages/auditcore_bpmn/src/auditcore_bpmn/citations.py
"""Normzitate erkennen und EU-Rechtsakte auflösen (CELEX, ELI, EUR-Lex-URL).

Rein syntaktisch, ohne Netzwerkzugriff. Erkannt werden die ausgeschriebene
Form („Artikel 73 Absatz 2 Buchstabe b der Verordnung (EU) 2021/1060“), die
Kurzform („Art. 74 Abs. 2 UAbs. 2“), Paragraphen („§ 44 Absatz 1 LHO“,
„§§ 23 und 44 LHO“), Verwaltungsvorschriften („VV Nummer 4.2 zu § 44 LHO“)
und Anhänge. Ergebnisse sind Vorschläge; übernommen werden sie durch die Anwendung.
"""

from __future__ import annotations

import re
from collections.abc import Callable, Iterator
from dataclasses import dataclass, replace
from typing import Protocol

from .extensions import LegalBasis, act_long


@dataclass(frozen=True)
class NormResolution:
    """Ergebnis der Auflösung eines Normnamens."""

    act: str
    celex: str | None = None
    eli: str | None = None
    url: str | None = None
    short_title: str | None = None
    document_type: str | None = None
    programming_period: str | None = None
    source: str = "syntaktisch"
    catalogued: bool = False


class NormResolver(Protocol):
    """Port zur Normauflösung (z. B. :class:`auditcore_bpmn.legal.LegalSourcesNormResolver`)."""

    def resolve(self, act: str) -> NormResolution | None:
        """Auflösung eines Normnamens oder ``None``."""


_EU_ACT = re.compile(
    r"(?P<kind>Delegierte(?:n)? Verordnung|Durchführungsverordnung|Verordnung|Richtlinie|Beschluss|Entscheidung)"
    r"\s*\((?:EU|EG|EWG|EU,\s*Euratom|EG,\s*Euratom)\)\s*(?P<nr>Nr\.\s*)?(?P<a>\d{1,4})/(?P<b>\d{1,4})"
)
_OLD_DIRECTIVE = re.compile(r"Richtlinie\s+(?P<a>\d{4})/(?P<b>\d{1,4})/(?:EU|EG|EWG)")
_ELI_KIND = {
    "Verordnung": "reg",
    "Delegierte Verordnung": "reg_del",
    "Delegierten Verordnung": "reg_del",
    "Durchführungsverordnung": "reg_impl",
    "Richtlinie": "dir",
    "Beschluss": "dec",
    "Entscheidung": "dec",
}
_CELEX_TYPE = {"reg": "R", "reg_del": "R", "reg_impl": "R", "dir": "L", "dec": "D"}


def _is_year(value: int) -> bool:
    return 1950 <= value <= 2100


def eu_act(act: str) -> tuple[str, int, int] | None:
    """``(ELI-Art, Jahr, Nummer)`` eines EU-Rechtsakts oder ``None``.

    Ab 2015 lautet die Zählung ``Jahr/Nummer``, davor ``Nr. Nummer/Jahr``;
    Richtlinien stets ``Jahr/Nummer``.
    """
    text = act_long(act)
    match = _EU_ACT.search(text)
    if match:
        kind, first, second = _ELI_KIND[match["kind"]], int(match["a"]), int(match["b"])
        number_first = kind != "dir" and (match["nr"] or (_is_year(second) and not _is_year(first)))
        year, number = (second, first) if number_first else (first, second)
        return (kind, year, number) if _is_year(year) else None
    match = _OLD_DIRECTIVE.search(text)
    return ("dir", int(match["a"]), int(match["b"])) if match else None


class EuActResolver:
    """Syntaktische Auflösung von EU-Rechtsakten ohne Katalog."""

    def resolve(self, act: str) -> NormResolution | None:
        """CELEX, ELI und EUR-Lex-URL eines EU-Rechtsakts oder ``None``."""
        found = eu_act(act)
        if found is None:
            return None
        kind, year, number = found
        celex = f"3{year}{_CELEX_TYPE[kind]}{number:04d}"
        return NormResolution(
            act=act_long(act),
            celex=celex,
            eli=f"http://data.europa.eu/eli/{kind}/{year}/{number}/oj",
            url=f"https://eur-lex.europa.eu/legal-content/DE/TXT/?uri=CELEX:{celex}",
        )


def enrich(legal_basis: LegalBasis, resolver: NormResolver | None = None) -> LegalBasis:
    """Füllt fehlende CELEX-/ELI-/URL-/Kurzbezeichnung; vorhandene Angaben bleiben."""
    if not legal_basis.act:
        return legal_basis
    result = (resolver or EuActResolver()).resolve(legal_basis.act)
    if result is None:
        return legal_basis
    return replace(
        legal_basis,
        celex=legal_basis.celex or result.celex,
        eli=legal_basis.eli or result.eli,
        url=legal_basis.url or result.url,
        short_title=legal_basis.short_title or result.short_title,
    )


# ---------------------------------------------------------------------------
# Zitate in Freitext
# ---------------------------------------------------------------------------

_EU_ACT_TEXT = (
    r"(?:(?:Delegierten|Durchführungs)\s*)?(?:Verordnung|Richtlinie|VO|RL)\s*"
    r"\((?:EU|EG|EWG|EU,\s*Euratom|EG,\s*Euratom)\)\s*(?:Nr\.\s*)?\d{1,4}/\d{1,4}"
)
_LONG_PARTS = (
    r"(?:\s+(?:Absatz|Abs\.)\s+(?P<abs>\d+[a-z]?))?(?:\s+Unterabsatz\s+(?P<uabs>\d+))?"
    r"(?:\s+(?:Satz|S\.)\s+(?P<satz>\d+))?(?:\s+Buchstabe\s+(?P<buchst>[a-z]{1,3}))?"
    r"(?:\s+(?:Nummer|Nr\.|Ziffer)\s+(?P<nr>[\divx.]+))?"
)
_SHORT_PARTS = (
    r"(?:\s*Abs\.\s*(?P<abs>\d+[a-z]?))?(?:\s*UAbs\.\s*(?P<uabs>\d+))?"
    r"(?:\s*(?:S\.|Satz)\s*(?P<satz>\d+))?(?:\s*(?:Buchst\.|lit\.)\s*(?P<buchst>[a-z]{1,3}))?"
    r"(?:\s*(?:Nr\.|Ziff\.)\s*(?P<nr>[\divx.]+))?"
)
_NATIONAL_ACT = r"(?P<norm>(?:[A-ZÄÖÜ][A-Za-zÄÖÜäöüß]*[A-ZÄÖÜ][A-Za-zÄÖÜäöüß]*|UVgO|VgV|GWB))\b"
_LONG = re.compile(
    r"(?:Artikel\s+(?P<art>\d+[a-z]?)|Anhang\s+(?P<anh>[IVXLC]+))"
    + _LONG_PARTS
    + r"\s+(?:der|des)\s+(?P<norm>"
    + _EU_ACT_TEXT
    + r")"
)
_SHORT = re.compile(
    r"Art\.\s*(?P<art>\d+[a-z]?)" + _SHORT_PARTS + r"(?:\s+(?P<norm>" + _EU_ACT_TEXT + r"|CPR|Dach-VO))?"
)
_SECTION = re.compile(r"§\s*(?P<par>\d+[a-z]?)" + _LONG_PARTS + r"\s+" + _NATIONAL_ACT)
_SECTIONS = re.compile(r"§§\s*(?P<liste>\d+[a-z]?(?:\s*(?:,|und|bis)\s*\d+[a-z]?)+)\s+" + _NATIONAL_ACT)
_ADMIN_RULE = re.compile(r"VV\s+(?:Nummer|Nr\.)\s*(?P<nr>[\d.]+)\s+zu\s+§\s*(?P<par>\d+[a-z]?)\s+" + _NATIONAL_ACT)


@dataclass(frozen=True)
class CitationMatch:
    """Ein erkanntes Zitat mit Fundstelle im Text."""

    legal_basis: LegalBasis
    start: int
    end: int
    text: str
    form: str


def _group(match: re.Match[str], name: str) -> str | None:
    try:
        value = match.group(name)
    except IndexError:
        return None
    return value.strip() if value else None


def _legal_basis(match: re.Match[str], act: str | None, **anchor: str | None) -> LegalBasis:
    """Rechtsgrundlage mit Untergliederung aus den benannten Gruppen des Treffers."""
    return LegalBasis(
        act=act,
        article=anchor.get("article"),
        section=anchor.get("section"),
        annex=anchor.get("annex"),
        paragraph=_group(match, "abs"),
        subparagraph=_group(match, "uabs"),
        sentence=_group(match, "satz"),
        point=_group(match, "buchst"),
        number=_group(match, "nr"),
    )


def _admin_rule(match: re.Match[str]) -> list[LegalBasis]:
    return [LegalBasis(act=f"VV zu § {match['par']} {match['norm']}", number=match["nr"])]


def _long(match: re.Match[str]) -> list[LegalBasis]:
    return [_legal_basis(match, match["norm"], article=_group(match, "art"), annex=_group(match, "anh"))]


def _short(match: re.Match[str]) -> list[LegalBasis]:
    return [_legal_basis(match, _group(match, "norm"), article=match["art"])]


def _sections(match: re.Match[str]) -> list[LegalBasis]:
    return [LegalBasis(act=match["norm"], section=number) for number in re.findall(r"\d+[a-z]?", match["liste"])]


def _section(match: re.Match[str]) -> list[LegalBasis]:
    return [_legal_basis(match, match["norm"], section=match["par"])]


#: Reihenfolge = Vorrang bei Überlappung.
_PATTERNS: tuple[tuple[str, re.Pattern[str], Callable[[re.Match[str]], list[LegalBasis]]], ...] = (
    ("vv", _ADMIN_RULE, _admin_rule),
    ("lang", _LONG, _long),
    ("kurz", _SHORT, _short),
    ("paragraphen", _SECTIONS, _sections),
    ("paragraph", _SECTION, _section),
)


def _matches(text: str) -> Iterator[CitationMatch]:
    taken: list[tuple[int, int]] = []
    for form, pattern, build in _PATTERNS:
        for match in pattern.finditer(text):
            span = (match.start(), match.end())
            if any(span[1] > start and span[0] < end for start, end in taken):
                continue
            taken.append(span)
            for legal_basis in build(match):
                yield CitationMatch(legal_basis.normalized(), span[0], span[1], match.group(0), form)


def find_citations(text: str | None) -> list[CitationMatch]:
    """Normzitate in Freitext (z. B. ``bpmn:documentation``), überlappungsfrei, nach Position."""
    if not text:
        return []
    return sorted(_matches(text), key=lambda hit: (hit.start, hit.legal_basis.citation()))


def parse_citation(text: str) -> LegalBasis:
    """Ein einzelnes Zitat strukturiert lesen; ohne eindeutigen Treffer bleibt es Freitext."""
    hits = find_citations(text)
    if len(hits) == 1 and hits[0].text.strip() == text.strip():
        return hits[0].legal_basis
    return LegalBasis(text=text.strip() or None)
